File: mnist.py
#!/bin/env python3
import numpy
import torch
import torchvision


class ANN(torch.nn.Module):

    def __init__(self):
        super(ANN, self).__init__()
        n_hidden = 128
        self.fc1 = torch.nn.Linear(28 * 28, n_hidden)
        self.fc2 = torch.nn.Linear(n_hidden, 10)

    def forward(self, x):
        x = torch.flatten(x, 1)
        x = self.fc1(x)
        x = torch.nn.functional.relu(x)
        x = self.fc2(x)
        x = torch.nn.functional.log_softmax(x, dim=1)
        return x

    def __str__(self):
        return 'ANN'


class DNN(torch.nn.Module):

    def __init__(self):
        super(DNN, self).__init__()
        self.conv1 = torch.nn.Conv2d(1, 32, 3, 1)
        self.conv2 = torch.nn.Conv2d(32, 64, 3, 1)
        self.dropout1 = torch.nn.Dropout(0.25)
        self.dropout2 = torch.nn.Dropout(0.5)
        self.fc1 = torch.nn.Linear(9216, 128)
        self.fc2 = torch.nn.Linear(128, 10)

    def forward(self, x):
        x = self.conv1(x)
        x = torch.nn.functional.relu(x)
        x = self.conv2(x)
        x = torch.nn.functional.relu(x)
        x = torch.nn.functional.max_pool2d(x, 2)
        x = self.dropout1(x)
        x = torch.flatten(x, 1)
        x = self.fc1(x)
        x = torch.nn.functional.relu(x)
        x = self.dropout2(x)
        x = self.fc2(x)
        x = torch.nn.functional.log_softmax(x, dim=1)
        return x

    def __str__(self):
        return 'DNN'


def norm(data, mean=None, std=None):
    data = data.double() / 255.0
    if mean is None:
        mean = data.sum() / numpy.array(data.shape).prod()
    if std is None:
        std = data.std()
    data = (data - mean) / std
    return data, mean, std


def train(model, device, data_train, optimizer):
    model.train()
    for data, label in data_train:
        optimizer.zero_grad()

        data = data.to(device)
        label = label.to(device)

        pred = model(data)
        loss = torch.nn.functional.nll_loss(pred, label)
        loss.backward()

        optimizer.step()


def test(model, device, data_test):
    model.eval()
    total_loss = 0.0
    num_correct = 0
    with torch.no_grad():
        for data, label in data_test:
            data = data.to(device)
            label = label.to(device)

            pred = model(data)
            loss = torch.nn.functional.nll_loss(pred, label).sum().item()
            total_loss += loss

            if label.item() == pred.argmax(keepdim=True).item():
                num_correct += 1

    print(
        f'loss {total_loss / len(data_test)} accuracy {num_correct / len(data_test)}'
    )


# create device
device = torch.device('cuda') if torch.cuda.is_available() else torch.device(
    'cpu')

# create datasets
transform = torchvision.transforms.Compose([
    torchvision.transforms.ToTensor(),
    torchvision.transforms.Normalize((0.1307, ), (0.3081, ))
])
data_train = torchvision.datasets.MNIST('data',
                                        train=True,
                                        download=True,
                                        transform=transform)
data_test = torchvision.datasets.MNIST('data',
                                       train=False,
                                       download=True,
                                       transform=transform)

# create data loader
data_valid = torch.utils.data.DataLoader(data_train)
data_train = torch.utils.data.DataLoader(data_train, 16)
data_test = torch.utils.data.DataLoader(data_test)

# create model and optimizer
model = ANN().to(device)
optimizer = torch.optim.Adadelta(model.parameters(), lr=0.005)
scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1, gamma=0.95)

# train
test(model, device, data_test)
for epoch in range(10):
    train(model, device, data_train, optimizer)
    test(model, device, data_test)
    scheduler.step()

    torch.save(model.state_dict(), f'mnist_{model}_{epoch}.pt')

File: mnist.cpp
#include <torch/torch.h>

#include <cstdint>
#include <iostream>
#include <string>

struct ANNImpl : public torch::nn::Module {
 public:
  ANNImpl(const int32_t num_hidden = 128)
      : fc1_(28 * 28, num_hidden), fc2_(num_hidden, 10) {
    register_module("fc1", fc1_);
    register_module("fc2", fc2_);
  }

  torch::Tensor forward(const torch::Tensor& in) {
    torch::Tensor x = in;
    x = torch::flatten(x, 1);
    x = fc1_->forward(x);
    x = torch::relu(x);
    x = fc2_->forward(x);
    x = torch::log_softmax(x, 1);

    return x;
  }

 private:
  torch::nn::Linear fc1_;
  torch::nn::Linear fc2_;
};

TORCH_MODULE(ANN);

struct DNNImpl : public torch::nn::Module {
 public:
  DNNImpl(const int32_t num_hidden = 128)
      : conv1_(torch::nn::Conv2dOptions(1, 32, 3)),
        conv2_(torch::nn::Conv2dOptions(32, 64, 3)),
        dropout1_(0.25),
        dropout2_(0.5),
        fc1_(9216, num_hidden),
        fc2_(num_hidden, 10) {
    register_module("conv1", conv1_);
    register_module("conv2", conv2_);
    register_module("dropout1", dropout1_);
    register_module("dropout2", dropout2_);
    register_module("fc1", fc1_);
    register_module("fc2", fc2_);
  }

  torch::Tensor forward(const torch::Tensor& in) {
    torch::Tensor x = in;

    x = conv1_->forward(x);
    x = torch::relu(x);
    x = conv2_->forward(x);
    x = torch::relu(x);
    x = torch::max_pool2d(x, 2);
    x = dropout1_->forward(x);
    x = torch::flatten(x, 1);
    x = fc1_->forward(x);
    x = torch::relu(x);
    x = dropout2_->forward(x);
    x = fc2_->forward(x);
    x = torch::log_softmax(x, 1);

    return x;
  }

 private:
  torch::nn::Conv2d conv1_;
  torch::nn::Conv2d conv2_;
  torch::nn::Dropout dropout1_;
  torch::nn::Dropout dropout2_;
  torch::nn::Linear fc1_;
  torch::nn::Linear fc2_;
};

TORCH_MODULE(DNN);

template <typename Model, typename DataLoader>
void train(Model& model, const torch::Device device, DataLoader& data_loader,
           torch::optim::Optimizer& optimizer) {
  model.train();

  for (auto& batch : data_loader) {
    optimizer.zero_grad();

    const torch::Tensor data{batch.data.to(device)};
    const torch::Tensor label{batch.target.to(device)};

    const torch::Tensor out{model.forward(data)};

    torch::Tensor loss{torch::nll_loss(out, label)};

    loss.backward();

    optimizer.step();
  }
}

template <typename Model, typename DataLoader>
void test(Model& model, const torch::Device device, DataLoader& data_loader) {
  torch::NoGradGuard no_grad;
  model.eval();

  double total_loss{0.0};
  int32_t num_correct{0};

  for (const auto& batch : data_loader) {
    const torch::Tensor data{batch.data.to(device)};
    const torch::Tensor label{batch.target.to(device)};

    const torch::Tensor out{model.forward(data)};

    const torch::Tensor loss{torch::nll_loss(out, label)};

    total_loss += loss.sum().item().to<double>();

    if (out.argmax(1).item().to<int32_t>() == label.item().to<int32_t>()) {
      ++num_correct;
    }
  }

  std::cout << "loss " << total_loss << " #correct " << num_correct
            << std::endl;
}

int main() {
  const torch::Device device{torch::cuda::is_available() ? torch::kCUDA
                                                         : torch::kCPU};

  const std::string path_data{"../data/MNIST/raw"};
  constexpr double mean{0.1307};
  constexpr double std{0.3081};
  const torch::data::transforms::Normalize<> transform(mean, std);
  auto train_data{torch::data::datasets::MNIST(path_data).map(transform).map(
      torch::data::transforms::Stack<>())};
  auto test_data{torch::data::datasets::MNIST(
                     path_data, torch::data::datasets::MNIST::Mode::kTest)
                     .map(transform)
                     .map(torch::data::transforms::Stack<>())};
  auto train_loader{torch::data::make_data_loader(train_data, 16)};
  auto test_loader{torch::data::make_data_loader(test_data, 1)};

  DNN model;
  model->to(device);

  torch::optim::SGD optimizer{model->parameters(),
                              torch::optim::SGDOptions(0.005).momentum(0.95)};

  test(*model, device, *test_loader);
  for (int32_t epoch = 0; epoch < 10; ++epoch) {
    train(*model, device, *train_loader, optimizer);
    test(*model, device, *test_loader);
  }

  torch::save(model, "mnist_model.pt");

  return EXIT_SUCCESS;
}